packages/contracts-bedrock/src/L2/SuperchainERC20Factory.sol

```diff
@@ -0,0 +1,186 @@
+        string memory _superSymbol = string.concat("S-", _symbol);
+
+        return _deploy(_remoteToken, _decimals, _superName, _superSymbol);
+    }
+
+    /// @notice Deploys a SuperchainERC20 from another L2.
+    /// @param _remoteToken Address of the remote token.
+    /// @param _decimals Decimals of the SuperchainERC20.
+    /// @param _name Name of the SuperchainERC20.
+    /// @param _symbol Symbol of the SuperchainERC20.
+    /// @return Address of the SuperchainERC20 deployment.
+    function deployFromL2(
+        address _remoteToken,
+        uint8 _decimals,
+        string memory _name,
+        string memory _symbol
+    )
+        external
+        returns (address)
+    {
+        if (msg.sender != L2_TO_L2_CROSS_DOMAIN_MESSENGER) revert CallerNotL2CrossDomainMessenger();
+
+        if (IL2ToL2CrossDomainMessenger(L2_TO_L2_CROSS_DOMAIN_MESSENGER).crossDomainMessageSender() != address(this)) {
+            revert InvalidCrossDomainSender();
+        }
+
+        return _deploy(_remoteToken, _decimals, _name, _symbol);
+    }
+
+    /// @notice Deploys a SuperchainERC20 Beacon Proxy using CREATE3.
+    /// @param _remoteToken Address of the remote token.
+    /// @param _decimals Decimals of the SuperchainERC20.
+    /// @param _name Name of the SuperchainERC20.
+    /// @param _symbol Symbol of the SuperchainERC20.
+    /// @return _superchainERC20 Address of the SuperchainERC20 deployment.
+    function _deploy(
+        address _remoteToken,
+        uint8 _decimals,
+        string memory _name,
+        string memory _symbol
+    )
+        internal
+        returns (address _superchainERC20)
+    {
+        // Revert if the deployment already exists
+        if (deploymentsData[_remoteToken].superchainERC20 != address(0)) revert AlreadyDeployed(_remoteToken);
+
+        // Encode the BeaconProxy creation code with the beacon contract address and metadata
+        bytes memory _creationCode = abi.encodePacked(
+            type(BeaconProxy).creationCode,
+            abi.encode(SUPERCHAIN_ERC20_BEACON, abi.encode(_remoteToken, _decimals, _name, _symbol))
+        );
+
+        // Use CREATE3 for deterministic deployment
+        bytes32 salt = keccak256(abi.encode(_remoteToken));
+        _superchainERC20 = CREATE3.deploy({ salt: salt, creationCode: _creationCode, value: 0 });
+
+        // Store SuperchainERC20 address and metadata
+        deploymentsData[_remoteToken] = DeploymentData(_superchainERC20, _decimals, _name, _symbol);
+
+        emit SuperchainERC20Deployed(_superchainERC20, _remoteToken, _decimals, _name, _symbol);
+    }
+
+    /// @notice Deploys a SuperchainERC20 from L2 with storage metadata.
+    /// @param _remoteToken Address of the remote token.
+    /// @param _chainIds Chain IDs to deploy to.
+    function deployWithMetadata(address _remoteToken, uint256[] memory _chainIds) external {
+        DeploymentData memory _metadata = deploymentsData[_remoteToken];
```

Comment:
fetching it from `storage` may be cheaper, we should test this, but this is very minor not priority